MTM-52486: document authentication with access token from external IAM (#962)
SoftwareAG/c8y-docs: content/users-guide/administration-bundle/configuring-single-sign-on.md

---

{{< product-c8y-iot >}} provides single sign-on (SSO) functionality, that allows a user to login with a single 3rd-party authorization server using the OAuth2 protocol, for example Azure Active Directory (ADD). Currently authorization code grant is supported only with access tokens in form of JWT.

> On top of the standard SSO, {{< product-c8y-iot >}} also allows you to access the platform resources using access tokens from your authorization server directly as Bearer token. More details on that in [Configuring authentication with OAuth2 access token from IAM](/users-guide/administration/#configuring-authentication-with-oauth2-access-token-from-iam).

Reviewer: ```suggestion
On top of the standard SSO, {{< product-c8y-iot >}} also allows you to access the platform resources using access tokens from your authorization server directly as Bearer token. For details refer to [Configuring authentication with OAuth2 access token from IAM](/users-guide/administration/#configuring-authentication-with-oauth2-access-token-from-iam).
```

> On top of the standard SSO, {{< product-c8y-iot >}} also allows you to access the platform resources using access tokens from your authorization server directly as Bearer token. For details refer to [Configuring authentication with OAuth2 access tokens from authorization server](/users-guide/administration/#configuring-authentication-with-oauth2-access-token-from-iam).

Reviewer: ```suggestion
On top of standard SSO, {{< product-c8y-iot >}} also allows you to access the platform resources using access tokens from your authorization server directly as a Bearer token. For details refer to [Configuring authentication with OAuth2 access tokens from authorization server](/users-guide/administration/#configuring-authentication-with-oauth2-access-token-from-iam).
```

Placeholders are not validated for correctness. Any not recognized or misspelled placeholder will be left in text unprocessed.
{{< /c8y-admon-info >}}

> ### Configuring authentication with OAuth2 access token from IAM

Reviewer: Have we already explained somewhere what IAM stands for? Otherwise I would mention it once, followed by IAM in brackets

Author: No I have not described it anywhere. I can also replace it, since I use IAM and authorization server interchangeably. 
There are many references of authorization server already.

> ### Configuring authentication with OAuth2 access tokens from authorization servers

Reviewer: Anchor is missing

> You can directly request {{< product-c8y-iot >}} using access tokens from your authorization server.

Reviewer: ```suggestion
You can directly request {{< product-c8y-iot >}} to use access tokens from your authorization server.
```

> This is suited for you, if you want your applications or users to access our resources without necessarily to need to log in to the platform

Reviewer: ```suggestion
This way, your applications or users can access our resources without the need to log in to the platform
```

or to use Basic authentication. With this, you can leverage your authorization server to get access tokens for your application which you will send in any subsequent request to
{{< product-c8y-iot >}}. 

{{< c8y-admon-req >}}
This feature requires the following on top of the above requirements:

* The authorization server you use supports OAuth2 client credentials grant type.
* All microservices are build with Microservice Java SDK version 1017.227.0 or higher. For custom-built microservices, refer to [General aspects > Security](/microservice-sdk/concept/#security) in the *Microservice SDK guide*.

{{< /c8y-admon-req >}}

> You can choose to enable or disable this authentication at your convenience under the **External token configuration** section.

Reviewer: ```suggestion
You can enable or disable this authentication option in the **External token configuration** section.
```

![External token disabled](/images/users-guide/Administration/sso-access-token-external-iam-disabled.png)
 

> If enabled, this authentication will take precedence over the standard [JWT token authentication](https://{{< domain-c8y >}}/guides/{{< c8y-current-version >}}/reference/rest-implementation/#jwt-token-authentication), which means for example that any HTTP request to {{< product-c8y-iot >}} with header `Authentication: Bearer {{access token}}` will assume the provenance of access token to be your IAM and not a {{< product-c8y-iot >}} issued access token.

Reviewer: ```suggestion
If enabled, this authentication will take precedence over the standard [JWT token authentication](https://{{< domain-c8y >}}/guides/{{< c8y-current-version >}}/reference/rest-implementation/#jwt-token-authentication), which means, for example, that any HTTP request to {{< product-c8y-iot >}} with the header `Authentication: Bearer {{access token}}` assumes that the provenance of the access token is your IAM instead of a {{< product-c8y-iot >}} issued access token.
```

> You will have to configure the user ID or the Application ID to any top-level claim in the access token.  

Reviewer: ```suggestion
You must configure the user ID or the application ID to any top-level claim in the access token.  
```

![External token user id](/images/users-guide/Administration/sso-access-token-external-iam-user-id-config.png)

> {{< product-c8y-iot >}} will create a user assigning it your configure user or app ID. Additionally, the user will be granted the roles and granted access to the applications defined in the **Access mapping** section.

Reviewer: ```suggestion
{{< product-c8y-iot >}} creates a user wich gets assigned the configured user ID or application ID. Additionally, this user will be granted the roles and access to the applications defined in the **Access mapping** section.
```

{{< c8y-admon-info >}}

> Notice that, in this configuration allows you to when set, to create a {{< product-c8y-iot >}} user representing your applications (the access tokens are obtained via the *client credentials flow*), 

Reviewer: ```suggestion
When set, this configuration allows you to create a {{< product-c8y-iot >}} user representing your applications (the access tokens are obtained via the *client credentials flow*), 
```

or your IAM users (the access tokens are obtained with the *password grant type*).
{{< /c8y-admon-info >}}

> By default, we verify that the token is not expired and its signature using the signature you have configured earlier.

Reviewer 2: I belive it rather should be "we" -> {{< product-c8y-iot >}}

> By default, we verify that the token is not expired and its signature using the signature you have configured earlier.

Reviewer: ```suggestion
By default, {{< product-c8y-iot >}} verifies that the token is not expired and its signature matches the signature you have configured earlier.
```

> You can strengthen the validation of the token by configuring either and introspection or a user info validation with the necessary credentials.

Reviewer: ```suggestion
You can strengthen the validation of the token by configuring either an introspection or a user info validation with the necessary credentials.
```

> They allow us to know whether the access tokens was purposefully invalidated or expired. In that case, using an invalidated access token you will be unauthorized to access {{< product-c8y-iot >}} resources. 

Reviewer: ```suggestion
This way, the platform knows if the access tokens were intentionally invalidated or expired. If you use an invalidated access token you will not be able to access {{< product-c8y-iot >}} resources. 
```

##### Introspection endpoint

> We use token introspection, to verify the validity of the access tokens of your applications. In general, this endpoint can be used for access tokens obtained via the client credentials flow or any other OAUTH2 flow.

Reviewer: ```suggestion
{{< product-c8y-iot >}} uses token introspection to verify the validity of the access tokens of your applications. In general, this endpoint can be used for access tokens obtained via the client credentials flow or any other OAuth2 flow.
```

> To configure the introspection, you must provide an introspection endpoint and a URL-encoded (x-form-urlencoded) body containing the access token, the client id and client secret, and a "Authorization" request header.  

Reviewer: ```suggestion
To configure the introspection you must provide an introspection endpoint and a URL-encoded (x-form-urlencoded) body containing the access token, the client ID and client secret, and a "Authorization" request header.  
```

> {{< product-c8y-iot >}} makes a request to the introspection endpoint of your authorization server to enquire about the status of the access token.

Reviewer: ```suggestion
{{< product-c8y-iot >}} makes a request to the introspection endpoint of your authorization server to enquire the status of the access token.
```

If the token is still active, then we proceed with verifying the token signature.

![External token introspection validation](/images/users-guide/Administration/sso-access-token-external-iam-introspection-validation.png)

> You can also decide to perform the introspection, for example every minute for the same access token by configuring the **Access token validation frequency** field.

Reviewer: ```suggestion
By configuring the **Access token validation frequency** accordingly, you can choose how frequently the introspection is performed 
```

> This is because it can be costly to always call the authorization server for the same access token. The default value is set to *1 minute*.

Reviewer: ```suggestion
as it may be costly to always call the authorization server for the same access token. The default value is set to *1 minute*.
```

##### User info endpoint

> The user info request can also be used to enquire on the validity of the access token of your users. 

Reviewer: ```suggestion
The user info request can also be used to enquire the validity of the access token of your users. 
```

> Unlike introspection, a user info request requires a user context. It means you cannot validate access tokens obtained with client credentials flow using this validation form, but instead you choose select the introspection validation.

Reviewer: ```suggestion
Unlike introspection, a user info request requires a user context. This means you cannot use this validation option to validate access tokens obtained with the client credentials flow, but must use the introspection validation instead.
```

Placeholders are not validated for correctness. Any not recognized or misspelled placeholder will be left in text unprocessed.
{{< /c8y-admon-info >}}

### Configuring authentication with OAuth2 access tokens from authorization servers

You can directly request {{< product-c8y-iot >}} to use access tokens from your authorization server.

> This way, your applications or users can access our resources without the need to log in to the platform

Reviewer 3: ```suggestion
This way, your applications or users can access resources without logging in to the platform
```

> or to use Basic authentication. With this, you can leverage your authorization server to get access tokens for your application which you will send in any subsequent request to

Reviewer 3: ```suggestion
or using Basic authentication. This leverages your authorization server to get access tokens for your applications which you can send in subsequent request to
```

Reviewer 3: "get" means "create"?

{{< product-c8y-iot >}}. 

{{< c8y-admon-req >}}
This feature requires the following on top of the above requirements:

> * The authorization server you use supports OAuth2 client credentials grant type.

Reviewer 3: ```suggestion
* Your authorization server must support the OAuth2 client credentials grant type.
```

* All microservices are build with Microservice Java SDK version 1017.227.0 or higher. For custom-built microservices, refer to [General aspects > Security](/microservice-sdk/concept/#security) in the *Microservice SDK guide*.

{{< /c8y-admon-req >}}

> You can enable or disable this authentication option in the **External token configuration** section.

Reviewer 3: ```suggestion
Enable or disable this authentication option in the **External token configuration** section.
```

![External token disabled](/images/users-guide/Administration/sso-access-token-external-iam-disabled.png)
 

> If enabled, this authentication will take precedence over the standard [JWT token authentication](https://{{< domain-c8y >}}/guides/{{< c8y-current-version >}}/reference/rest-implementation/#jwt-token-authentication), which means, for example, that any HTTP request to {{< product-c8y-iot >}} with the header `Authentication: Bearer {{access token}}` assumes that the provenance of the access token is your authorization server instead of a {{< product-c8y-iot >}} issued access token.

Reviewer 3: ```suggestion
If enabled, this authentication takes precedence over the standard [JWT token authentication](https://{{< domain-c8y >}}/guides/{{< c8y-current-version >}}/reference/rest-implementation/#jwt-token-authentication), which means, for example, that an HTTP request to {{< product-c8y-iot >}} with the header `Authentication: Bearer {{access token}}` assumes that the source of the access token is your authorization server instead of the token being issued by {{< product-c8y-iot >}}.
```

> You must configure the user ID or the application ID to any top-level claim in the access token.

Reviewer 3: ```suggestion
Configure the user ID or the application ID to any top-level claim in the access token.
```

![External token user id](/images/users-guide/Administration/sso-access-token-external-iam-user-id-config.png)

> {{< product-c8y-iot >}} creates a user wich gets assigned the configured user ID or application ID. Additionally, this user will be granted the roles and access to the applications defined in the **Access mapping** section.

Reviewer 3: ```suggestion
{{< product-c8y-iot >}} creates a user wich gets assigned the configured user ID or application ID. Additionally, this user is granted the roles to access to the applications defined in the **Access mapping** section.
```

{{< c8y-admon-info >}}

> When set, this configuration allows you to create a {{< product-c8y-iot >}} user representing your applications (the access tokens are obtained via the *client credentials flow*), 
> or the users of your authorization server (the access tokens are obtained with the *password grant type*).

Reviewer 3: ```suggestion
If it is set, the configuration allows you to create a {{< product-c8y-iot >}} user representing your applications (the access tokens are obtained via the *client credentials flow*), 
or the users of your authorization server (the access tokens are obtained with the *password grant type*).
```

{{< /c8y-admon-info >}}

By default, {{< product-c8y-iot >}} verifies that the token is not expired and its signature matches the signature you have configured earlier.
You can strengthen the validation of the token by configuring either an introspection or a user info validation with the necessary credentials.

> This way, the platform knows if the access tokens were intentionally invalidated or expired. If you use an invalidated access token you will not be able to access {{< product-c8y-iot >}} resources. 

Reviewer 3: ```suggestion
This way, the platform knows if the access tokens were intentionally invalidated or expired. You cannot access {{< product-c8y-iot >}} resources with an invalidated access token. 
```

##### Introspection endpoint

{{< product-c8y-iot >}} uses token introspection to verify the validity of the access tokens of your applications. In general, this endpoint can be used for access tokens obtained via the client credentials flow or any other OAuth2 flow.

To configure the introspection you must provide an introspection endpoint and a URL-encoded (x-form-urlencoded) body containing the access token, the client ID and client secret, and a "Authorization" request header.  

> {{< product-c8y-iot >}} makes a request to the introspection endpoint of your authorization server to enquire the status of the access token.

Reviewer 3: ```suggestion
{{< product-c8y-iot >}} requests the introspection endpoint of your authorization server to query the status of the access token.
```

> If the token is still active, then we proceed with verifying the token signature.

Reviewer 3: ```suggestion
If the token is active, proceed with verifying the token signature.
```

![External token introspection validation](/images/users-guide/Administration/sso-access-token-external-iam-introspection-validation.png)

> By configuring the **Access token validation frequency** accordingly, you can choose how frequently the introspection is performed 

Reviewer 3: ```suggestion
You can configure the **Access token validation frequency** to set how often the introspection is performed 
```

> as it may be costly to always call the authorization server for the same access token. This implies that the validation status for this token is cached internally for that amount of time specified.

Reviewer 3: ```suggestion
as it may be costly to always call the authorization server for the same access token. The validation status of the token is cached internally for the specified time.
```

If the token is revoked in the meantime, {{< product-c8y-iot >}} will only be aware during the next validation, i.e. the token will be considered still for the remaining of time until the next validation.

> For this reason, you should use a frequency. The default value is set to *1 minute*.

Reviewer 3: ```suggestion
To avoid this, use a frequency. The default value is one minute.
```

![External token validation interval](/images/users-guide/Administration/sso-access-token-external-iam-validation-interval.png)

##### User info endpoint

> The user info request can also be used to enquire the validity of the access token of your users. 

Reviewer 3: ```suggestion
The user info request can also be used to check the validity of the access token of your users. 
```

> Unlike introspection, a user info request requires a user context. This means you cannot use this validation option to validate access tokens obtained with the client credentials flow, but must use the introspection validation instead.

Reviewer 3: ```suggestion
Unlike introspection, a user info request requires a user context. This means you cannot use it to validate access tokens obtained with the client credentials flow.
```

![External token userinfo validation](/images/users-guide/Administration/sso-access-token-external-iam-userinfo-validation.png)

{{< c8y-admon-caution >}}

>  * If you decide to use either of the two validations, you need to make sure that your authorization server exposes the introspection or the user info endpoint.

Reviewer 3: ```suggestion
If you use either of the two validations, make sure that your authorization server exposes the introspection or the user info endpoint.
```
